esp-wifi/src/wifi/mod.rs

```diff
@@ -340,6 +349,192 @@ impl Default for ClientConfiguration {
+            channel_filter_en: true,
+            manu_scale: false,
+            shift: 0,
+            dump_ack_en: false,
+        }
+    }
+
+    #[cfg(esp32c6)]
+    fn default() -> Self {
+        // https://github.com/esp-rs/esp-wifi-sys/blob/2a466d96fe8119d49852fc794aea0216b106ba7b/esp-wifi-sys/headers/esp_wifi_he_types.h#L67-L82
+        Self {
+            enable: 1,
+            acquire_csi_legacy: 1,
+            acquire_csi_ht20: 1,
+            acquire_csi_ht40: 1,
+            acquire_csi_su: 1,
+            acquire_csi_mu: 1,
+            acquire_csi_dcm: 1,
+            acquire_csi_beamformed: 1,
+            acquire_csi_he_stbc: 2,
+            val_scale_cfg: 2,
+            dump_ack_en: 1,
+            reserved: 19,
+        }
+    }
+}
+
+#[cfg(csi_enable)]
+impl From<CsiConfig> for wifi_csi_config_t {
+    fn from(config: CsiConfig) -> Self {
+        #[cfg(not(esp32c6))]
+        {
+            wifi_csi_config_t {
+                lltf_en: config.lltf_en,
+                htltf_en: config.htltf_en,
+                stbc_htltf2_en: config.stbc_htltf2_en,
+                ltf_merge_en: config.ltf_merge_en,
+                channel_filter_en: config.channel_filter_en,
+                manu_scale: config.manu_scale,
+                shift: config.shift,
+                dump_ack_en: config.dump_ack_en,
+            }
+        }
+        #[cfg(esp32c6)]
+        {
+            wifi_csi_acquire_config_t {
+                _bitfield_align_1: [0; 0],
+                _bitfield_1: wifi_csi_acquire_config_t::new_bitfield_1(
+                    config.enable,
+                    config.acquire_csi_legacy,
+                    config.acquire_csi_ht20,
+                    config.acquire_csi_ht40,
+                    config.acquire_csi_su,
+                    config.acquire_csi_mu,
+                    config.acquire_csi_dcm,
+                    config.acquire_csi_beamformed,
+                    config.acquire_csi_he_stbc,
+                    config.val_scale_cfg,
+                    config.dump_ack_en,
+                    config.reserved,
+                ),
+            }
+        }
+    }
+}
+
+#[cfg(csi_enable)]
+impl CsiConfig {
+    /// Set CSI data configuration
+    pub(crate) fn apply_config(&self) -> Result<(), WifiError> {
+        // let conf = self.clone().into();
+        let conf: wifi_csi_config_t = self.clone().into();
+
+        unsafe {
+            esp_wifi_result!(esp_wifi_set_csi_config(&conf))?;
+        }
+        Ok(())
+    }
+
+    /// Register the RX callback function of CSI data. Each time a CSI data is
+    /// received, the callback function will be called.
+    pub(crate) fn set_receive_cb(
+        &mut self,
+        cb: fn(crate::binary::include::wifi_csi_info_t),
+    ) -> Result<(), WifiError> {
+        unsafe {
+            esp_wifi_result!(esp_wifi_set_csi_rx_cb(
+                Some(promiscuous_csi_rx_cb),
+                cb as *mut crate::wifi::c_types::c_void
```

Comment:
I think we might have to `Box` this callback, or maybe associate some lifetime to avoid pointing to a callback that no longer exists in memory. Especially if we address the other review comment about changing the API to take a closure instead of a function pointer.

Imagine the following scenario:

```rust
fn install_callback(controller: &mut WifiController) {
	let my_closure = |data| { /* do stuff with csi data*/}; // closure allocated on the stack
	controller.set_csi(Default::default(), my_closure).unwrap();
}
// after calling install_callback the callback pointer points to uninitialized memory :(
```

It's probably possible to do the same with a function pointer, just less likely than with a closure.

Reply:
I'd say a plain-old function pointer won't get invalidated but yes - a closure might not be valid forever